Kernel/FileSystem/ISO9660FileSystem.cpp

```diff
@@ -0,0 +1,625 @@
+        return KSuccess;
+    }
+
+    ISO9660DirectoryIterator iterator { const_cast<ISO9660FS&>(*this), record };
+
+    while (!iterator.done()) {
+        auto maybe_result = visitor(*iterator);
+        if (maybe_result.is_error()) {
+            return maybe_result.error();
+        }
+
+        auto result = maybe_result.value();
+        if (result == RecursionDecision::Recurse) {
+            auto maybe_has_moved = iterator.next();
+            if (maybe_has_moved.is_error()) {
+                return maybe_has_moved.error();
+            }
+
+            if (!maybe_has_moved.value()) {
+                // If next() hasn't moved then we have read through all the
+                // directories, and can exit.
+                return KSuccess;
+            }
+
+            continue;
+        } else if (result == RecursionDecision::Break) {
+            return KSuccess;
+        } else {
+            while (!iterator.done()) {
+                if (!iterator.skip()) {
+                    if (!iterator.go_up()) {
+                        return KSuccess;
+                    }
+                } else {
+                    break;
+                }
+            }
+        }
+    }
+
+    return KSuccess;
+}
+
+KResultOr<size_t> ISO9660Inode::read_bytes(off_t offset, size_t size, UserOrKernelBuffer& buffer, FileDescription*) const
+{
+    MutexLocker inode_locker(m_inode_lock);
+
+    auto& file_system = const_cast<ISO9660FS&>(static_cast<ISO9660FS const&>(fs()));
+    u32 data_length = LittleEndian { m_record.data_length.little };
+    u32 extent_location = LittleEndian { m_record.extent_location.little };
+
+    if (static_cast<u64>(offset) >= data_length)
+        return 0;
+
+    auto block = KBuffer::try_create_with_size(file_system.m_logical_block_size);
+    if (!block) {
+        return ENOMEM;
+    }
+    auto block_buffer = UserOrKernelBuffer::for_kernel_buffer(block->data());
+
+    size_t total_bytes = min(size, data_length - offset);
+    size_t nread = 0;
+    size_t blocks_already_read = offset / file_system.m_logical_block_size;
+    size_t initial_offset = offset % file_system.m_logical_block_size;
+
+    auto current_block_index = BlockBasedFileSystem::BlockIndex { extent_location + blocks_already_read };
+    while (nread != total_bytes) {
+        size_t bytes_to_read = min(total_bytes - nread, file_system.logical_block_size() - initial_offset);
+        auto buffer_offset = buffer.offset(nread);
+        dbgln_if(ISO9660_DEBUG, "ISO9660Inode::read_bytes: Reading {} bytes into buffer offset {}/{}, logical block index: {}", bytes_to_read, nread, total_bytes, current_block_index.value());
+
+        if (bool result = file_system.raw_read(current_block_index, block_buffer); !result) {
+            return EIO;
+        }
+
+        bool result = buffer_offset.write(block->data() + initial_offset, bytes_to_read);
+        if (!result) {
+            return EIO;
```

Comment:
The error here is when copying to the buffer fails due to a segfault.
```suggestion
            return EFAULT;
```